graphs/lca.cpp

```diff
@@ -1,35 +1,39 @@
+#include <bits/stdc++.h>
+using namespace std;
 struct lca {
+	using pii = pair<int, int>;
 	int Log;
 	vector<vector<int>> memo;
-	vector<int> depth;
+	vector<int> depth, dist;
 
-	lca(const vector<vector<int>> &graph, int root) {//0 - based nodes
+	// use weights of 1 for unweighted tree
+	lca(const vector<vector<pii>> &graph, int root = 0) {//0 - based nodes
 		int n = graph.size();
 		Log = 1;
 		while((1<<Log) < n) ++Log;
 		depth.resize(n,0);
+		dist.resize(n,0);
 		memo.resize(n,vector<int>(Log));
 		dfs(root,root,graph);
 	}
 
-	void dfs(int node, int par, const vector<vector<int>> &graph) {
+	void dfs(int node, int par, const vector<vector<pii>> &graph) {
 		memo[node][0] = par;
 		for(int i = 1; i < Log; ++i) {
 			memo[node][i] = memo[memo[node][i-1]][i-1];
 		}
-		for(int to : graph[node]) {
+		for(auto [to, w] : graph[node]) {
```

Comment:
loop by reference:

for(auto &[to ....

Reply:
Looping by reference does not make this faster, but I will do as you request as you may have other reasons. Looping by reference is the same number of instructions because we have to load the value from the address we are given in the loop or we have to load the value before the loop and keep a pointer to the value we are on (in which case we would load the value at every iteration from anyway).

Reply:
I see, didn't know this. It's fine to loop by value then